include/ginkgo/core/matrix/bccoo.hpp

```diff
@@ -0,0 +1,620 @@
+     * @param alpha  scaling of the result of Bccoo * b
+     * @param b  vector(s) on which the operator is applied
+     * @param x  output vector(s)
+     *
+     * @return this
+     */
+    LinOp* apply2(const LinOp* alpha, const LinOp* b, LinOp* x)
+    {
+        this->validate_application_parameters(b, x);
+        GKO_ASSERT_EQUAL_DIMENSIONS(alpha, dim<2>(1, 1));
+        auto exec = this->get_executor();
+        this->apply2_impl(make_temporary_clone(exec, alpha).get(),
+                          make_temporary_clone(exec, b).get(),
+                          make_temporary_clone(exec, x).get());
+        return this;
+    }
+
+    /**
+     * @copydoc apply2(const LinOp *, const LinOp *, LinOp *)
+     */
+    const LinOp* apply2(const LinOp* alpha, const LinOp* b, LinOp* x) const
+    {
+        this->validate_application_parameters(b, x);
+        GKO_ASSERT_EQUAL_DIMENSIONS(alpha, dim<2>(1, 1));
+        auto exec = this->get_executor();
+        this->apply2_impl(make_temporary_clone(exec, alpha).get(),
+                          make_temporary_clone(exec, b).get(),
+                          make_temporary_clone(exec, x).get());
+        return this;
+    }
+
+protected:
+    /**
+     * Creates an empty BCCOO matrix
+     *
+     * @param exec  Executor associated to the matrix
+     */
+    Bccoo(std::shared_ptr<const Executor> exec)
+        : EnableLinOp<Bccoo>(exec, dim<2>{}),
+          rows_(exec, 0),
+          cols_(exec, 0),
+          types_(exec, 0),
+          offsets_(exec, 0),
+          chunk_(exec, 0),
+          num_nonzeros_{0},
+          block_size_{0},
+          compression_{bccoo::compression::def_value}
+    {}
+
+    /**
+     * Creates an empty BCCOO matrix, fixing compression level and blocksize
+     *
+     * @param exec  Executor associated to the matrix
+     */
+    Bccoo(std::shared_ptr<const Executor> exec, index_type block_size,
+          bccoo::compression compression)
+        : EnableLinOp<Bccoo>(exec, dim<2>{}),
+          rows_(exec, 0),
+          cols_(exec, 0),
+          types_(exec, 0),
+          offsets_(exec, 0),
+          chunk_(exec, 0),
+          num_nonzeros_{0},
+          block_size_{block_size},
+          compression_{compression}
+    {}
+    /**
+     * Creates an uninitialized BCCOO matrix of the specified sizes.
+     *
+     * @param exec  Executor associated to the matrix
+     * @param size  size of the matrix
+     * @param num_nonzeros  number of nonzeros
+     * @param block_size    number of nonzeros in each block
+     * @param num_bytes     number of bytes
+     * @param compression   compression used in the definition
+     */
+    Bccoo(std::shared_ptr<const Executor> exec, const dim<2>& size,
+          index_type num_nonzeros, index_type block_size, size_type num_bytes,
+          bccoo::compression compression)
+        : EnableLinOp<Bccoo>(exec, size),
+          rows_(exec,
+                (block_size == 0) ? 0 : ceildiv(num_nonzeros, block_size)),
```

Comment:
for `ceildiv`, you need to include `ginkgo/core/base/math.hpp`

Reply:
Done!!